### PR TITLE
Add new `trivial_map_over_range` lint

### PR DESCRIPTION
**Pull request by rspencer01**
This lint checks for code that looks like
```rust
  let something : Vec<_> = (0..100).map(|_| {
    1 + 2 + 3
  }).collect();
```
which is more clear as
```rust
  let something : Vec<_> = std::iter::repeat_with(|| {
    1 + 2 + 3
  }).take(100).collect();
```

That is, a map over a range which does nothing with the parameter passed to it is simply a function (or closure) being called `n` times and could be more semantically expressed using `take`.

- [x] Followed [lint naming conventions][lint_naming]
- [x] Added passing UI tests (including committed `.stderr` file)
- [x] `cargo test` passes locally
- [x] Executed `cargo dev update_lints`
- [x] Added lint documentation
- [x] Run `cargo dev fmt`

changelog: new lint: [`trivial_map_over_range`] `restriction`

### CODE REVIEW

**Comment by rspencer01** [comment]
I would argue that if you are using a `DoubleEndedIterator` on a `map` that is independent of the input, you are probably not expecting the behaviour that you will get. On the other hand, I could see this being used in a `dyn impl` somewhere[^1]. Your point about other traits is also good.

I always have a mental hiccough when I see this pattern and have to think about _why_ it was written that way (does the author want to highlight that really we aren't taking `end-start` elements, but we are morally acting on something `[start..end]`? I know it is quite common though. I have a hunch that this is an attempt at rewriting `for _ in 0..n { ... }` in a more iterator-y way though (certainly that's why I did it in the past). This doesn't feel like it _should_ be idiomatic, even if it has become so.

At the end of the day though this is perhaps a slightly personal-stylistic and opinionated lint. @y21, would you consider this under a different category?

[^1]: Though it is sad that `Take<RepeatWith>` doesn't implement `DoubleEndedIterator`.

**Comment by y21** [comment]
The reason for why I brought up the missing `DoubleEndedIterator` impl is because this warns on [code in the regex crate](https://docs.rs/regex-automata/latest/src/regex_automata/nfa/thompson/compiler.rs.html#1330) where this lint's suggestion would not work.
<details>
<summary>Other warnings from a lintcheck run</summary>

```
target/lintcheck/sources/crossbeam-deque-0.8.5/src/deque.rs:42:43 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/crossbeam-utils-0.8.20/src/sync/sharded_lock.rs:104:110 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/rayon-1.10.0/src/iter/par_bridge.rs:89:89 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/rayon-core-1.12.1/src/broadcast/mod.rs:110:111 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/rayon-core-1.12.1/src/broadcast/mod.rs:141:147 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/rayon-core-1.12.1/src/registry.rs:241:251 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/rayon-core-1.12.1/src/registry.rs:254:259 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/rayon-core-1.12.1/src/scope/mod.rs:560:560 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/rayon-core-1.12.1/src/scope/mod.rs:653:656 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/rayon-core-1.12.1/src/sleep/mod.rs:64:64 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/regex-automata-0.4.7/src/nfa/thompson/compiler.rs:1330:1330 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/sharded-slab-0.1.7/src/shard.rs:97:97 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
target/lintcheck/sources/thread_local-1.1.8/src/lib.rs:511:515 clippy::trivial_map_over_range "map of a trivial closure (not dependent on parameter) over a range"
| clippy::trivial_map_over_range                     |    13 |
```
Most of them are true positives though, except for the mentioned regex case.
</details>

------------

For the category, I would personally put this in pedantic or even restriction because I still feel like this is such a common pattern (e.g. clippy itself has a suggestion that goes against this lint), but we can also wait until the FCP where other team members have a last look over this and we can see what others think. (It's entirely possible that more people disagree with my opinion :))

**Comment by rspencer01** [comment]
Thank you for your detailed comments. This is my first clippy lint and my first time working with things like `hir`, so apologies for the rough edges. I will look at and address your comments, move the lint to pedantic or restriction and then get back to you.

**Comment by y21** [comment]
No worries, I'll change the label here to clean up my queue, you can write `@rustbot ready` to change it back (or if you're stuck, feel free to to ask for help).

@rustbot author

File: tests/ui/trivial_map_over_range.stderr [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  --> tests/ui/trivial_map_over_range.rs:15:5
+   |
+LL |     (3..10).map(|_| do_something());
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `std::iter::repeat_with(|| do_something()).take(7)`
+
+error: map of a trivial closure (not dependent on parameter) over a range
+  --> tests/ui/trivial_map_over_range.rs:16:5
+   |
+LL |     (0..10).map(|_| 3);
+   |     ^^^^^^^^^^^^^^^^^^ help: use: `std::iter::repeat_with(|| 3).take(10)`
```
[the remark this answers is not in the document]
Comment (rspencer01) [reply tests/ui/trivial_map_over_range.stderr:20]: My hope for this was that it would get picked up by another lint looking for eager evaluable closures in `repeat_with`. I'm not sure what clippy's thoughts are regarding multi-step corrections: I'd believe that the lint should do as much as it can (i.e. suggest collapsing to `repeat` instead of `repeat_with`.

File: clippy_lints/src/trivial_map_over_range.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+use rustc_data_structures::packed::Pu128;
+use rustc_errors::Applicability;
+use rustc_hir::{Body, Closure, Expr, ExprKind, LangItem, PatKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Lint to prevent trivial `map`s over ranges when one could use `take`.
```
[the remark this answers is not in the document]
Comment (rspencer01) [reply clippy_lints/src/trivial_map_over_range.rs]: I agree and don't like the word `trivial` in this context any more. I've renamed the lint `map_with_unused_parameter_over_range` which is very verbose, but naming things is hard and any help would be appreciated.

File: clippy_lints/src/trivial_map_over_range.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    "map of a trivial closure (not dependent on parameter) over a range"
+}
+
+declare_lint_pass!(TrivialMapOverRange => [TRIVIAL_MAP_OVER_RANGE]);
+
+impl LateLintPass<'_> for TrivialMapOverRange {
+    fn check_expr(&mut self, cx: &LateContext<'_>, ex: &Expr<'_>) {
+        if let ExprKind::MethodCall(path, receiver, [map_arg_expr], _call_span) = ex.kind
+            && path.ident.name == rustc_span::sym::map
+            && let ExprKind::Struct(qpath, fields, _) = receiver.kind
```
[the remark this answers is not in the document]
Comment (rspencer01) [reply clippy_lints/src/trivial_map_over_range.rs]: Done and got a few freebees from it.

File: clippy_lints/src/trivial_map_over_range.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            && matches!(qpath, QPath::LangItem(LangItem::Range, _))
+            && fields.len() == 2
+            && let ExprKind::Closure(Closure { body, .. }) = map_arg_expr.kind
+            && let Body { params: [param], .. } = cx.tcx.hir().body(*body)
+            && matches!(param.pat.kind, PatKind::Wild)
+            && let ExprKind::Lit(lit) = fields[0].expr.kind
+            && let LitKind::Int(Pu128(lower_bound), _) = lit.node
```
[the remark this answers is not in the document]
Comment (rspencer01) [reply clippy_lints/src/trivial_map_over_range.rs]: Done

File: tests/ui/trivial_map_over_range.rs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    let upper = 4;
+    (0..upper).map(|_| do_something());
+    let upper_fn = || 4;
+    (0..upper_fn()).map(|_| do_something());
+    // These should not be raised
+    (0..=10).map(|_| do_something()); // Inclusive bounds not yet handled
+    (0..10).map(|x| do_something()); // We do not detect unused parameters
+    (0..10).map(|x| do_something()).collect::<Vec<_>>(); // We do not detect unused parameters
+    (0..10).map(|x| do_something_interesting(x, 4)); // Actual map over range
+    "Foobar".chars().map(|_| do_something()); // Not a map over range
```
[the remark this answers is not in the document]
Comment (rspencer01) [reply tests/ui/trivial_map_over_range.rs]: Done, though the latter fails. This would require more in-depth handling of ranges.

File: clippy_lints/src/trivial_map_over_range.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (y21) [reply clippy_lints/src/trivial_map_over_range.rs]: Yeah, it's a bit wordy but I also can't think of something better. Some similar lints follow the pattern `manual_*`, so e.g. `manual_repeat_take` but not sure it's much better or applicable here. We could also just leave it for the FCP and see if others have a better name

**Review by y21 — COMMENTED** [review]
Sorry for taking a bit longer to respond here, have been a bit busy recently 😓

File: clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        // `repeat`
+        let snippet =
+            snippet_with_applicability(cx, arg.span, "|| { ... }", &mut applicability).replacen("|_|", "||", 1);
+        span_lint_and_sugg(
+            cx,
+            MAP_WITH_UNUSED_ARGUMENT_OVER_RANGES,
+            ex.span,
+            "map of a closure that does not depend on its parameter over a range",
+            "use",
+            format!("std::iter::repeat_with({snippet}).take({count})"),
```
[the remark this answers is not in the document]
Comment (y21) [reply clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs:74]: also a more descriptive help message than "use" that makes sense on its own without the suggestion would be nice, since e.g. rust-analyzer will display the help message on its own without the suggestion in the quick fix list (this is a problem that a lot of existing lints also have FWIW)

File: clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                cx,
+                MAP_WITH_UNUSED_ARGUMENT_OVER_RANGES,
+                ex.span,
+                "map of a closure that does not depend on its parameter over a range",
+                |diag| {
+                    diag.multipart_suggestion(
+                        "remove the explicit range and use `repeat_with` and `take`",
+                        vec![
+                            (receiver.span.to(method_call_span), "std::iter::repeat_with".to_owned()),
+                            (param.span, String::new()),
```
Comment (y21) [inline clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs:109]: The only difference between this branch and the `span_lint_and_then` above is the function to suggest (repeat/repeat_with) and which span to remove right?
Can we just select those two based on `switch_to_eager_eval()` and merge all the rest of the `two span_lint_and_then calls` so the lint messages don't need to be repeated?

File: clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        && let LitKind::Int(Pu128(lower_bound), _) = lit.node
+    {
+        if let ExprKind::Lit(lit) = end.kind
+            && let LitKind::Int(Pu128(upper_bound), _) = lit.node
+        {
+            // Here we can explicitly calculate the number of iterations
+            let count = if upper_bound >= lower_bound {
+                match range.limits {
+                    RangeLimits::HalfOpen => upper_bound - lower_bound,
+                    RangeLimits::Closed => upper_bound - lower_bound + 1,
```
Comment (y21) [inline clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs:35]: I think `(0..=u128::MAX).map(...)` will still ICE because of overflow here. It does seem like a strange thing to write in the first place considering it will also panic at runtime, but would be good to avoid a crash there.

Maybe we should just ignore that case altogether with `.checked_add(1)?`

**Review by y21 — APPROVED** [review]
Three small things then I think this should be good :)

**Comment by y21** [comment]
The implementation looks good to me. I opened a thread on zulip for the FCP process that we require for new lints: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/FCP.3A.20map_with_unused_argument_over_ranges

Would be good if you could squash the commits, but that can also wait until after the FCP is complete since you might end up having to rebase anyway

**Comment by rspencer01** [comment]
- [ ] squash and rebase

**Comment by y21** [comment]
It's been a few weeks since the FCP has started. A few things have come up:

- The lint category should be restriction (see [the poll])
- Instead of `repeat(<element>).take(<count>)` we can suggest `repeat_n(<element>, <count>)` if the MSRV is >=1.83


[the poll]: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/FCP.3A.20map_with_unused_argument_over_ranges/near/469060597

**Comment by xFrednet** [comment]
Hey @rspencer01, this is a ping from triage, since there hasn't been any activity in some time. Have you seen the latest comment? Once that part is fixed, this should be good to go :D

If you have any questions, you're always welcome to ask them in this PR or on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy).

@rustbot author

**Comment by rspencer01** [comment]
Thanks, this is on my todo list! Sorry its taken me a while to get back to it. I hope to have fixes for the two comments soon.

**Comment by rspencer01** [comment]
Checks seem to be failing due to an unrelated `GLIBC` error.

@rustbot ready

**Comment by y21** [comment]
Hm, might be spurious? Let's see what's up with that error.
(I'd also like to see the lintcheck results before merging)

@bors try

File: clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+                            format!("std::iter::{method_to_use_name}"),
+                        ),
+                        new_span,
+                        (
+                            ex.span.shrink_to_hi(),
+                            if use_take {
+                                format!(".take({count})")
+                            } else {
+                                String::new()
+                            },
```
Comment (y21) [inline clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs:127]: These empty replacements (empty suggestion and empty span) trigger a [debug assertion in the compiler](https://github.com/rust-lang/rust/blob/e92993dbb43f0a5d17fe56e2d82f90435d6521c8/compiler/rustc_errors/src/diagnostic.rs#L922), so this needs to conditionally push the replacement into the vec

File: clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        if lower_bound == 0 {
+            if range.limits == RangeLimits::Closed {
+                return Some(format!("{end_snippet} + 1"));
+            }
+            return Some(end_snippet);
+        }
+        if range.limits == RangeLimits::Closed {
+            return Some(format!("{end_snippet} - {}", lower_bound - 1));
+        }
+        return Some(format!("{end_snippet} - {lower_bound}"));
```
Comment (y21) [inline clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs:55]: This is a semantic change in case of underflow, right? For x = 0, `(2..x).map(|_| ()).count()` saturates and returns 0 while the suggestion `repeat_n((), x - 2).count()` either panics in debug mode or returns `18446744073709551614`. Not sure how likely it is that it affects real code, but we might still want to adjust the applicability to MaybeIncorrect if we get here to be safe, since the user might need saturating_sub in rare cases?

File: clippy_lints/src/matches/single_match.rs [hunk 11]
```diff
@@ -249,7 +249,10 @@ impl<'a> PatState<'a> {
         let states = match self {
             Self::Wild => return None,
             Self::Other => {
-                *self = Self::StdEnum(cx.arena.alloc_from_iter((0..adt.variants().len()).map(|_| Self::Other)));
+                *self = Self::StdEnum(
+                    cx.arena
+                        .alloc_from_iter(std::iter::repeat_with(|| Self::Other).take(adt.variants().len())),
```
[the remark this answers is not in the document]
Comment (rspencer01) [reply clippy_lints/src/matches/single_match.rs:254]: Happy for me to implement (derive) clone for `PatState`?

File: clippy_config/src/msrvs.rs [hunk 12]
```diff
@@ -17,7 +17,7 @@ macro_rules! msrv_aliases {
 
 // names may refer to stabilized feature flags or library items
 msrv_aliases! {
-    1,83,0 { CONST_EXTERN_FN, CONST_FLOAT_BITS_CONV, CONST_FLOAT_CLASSIFY }
+    1,83,0 { CONST_EXTERN_FN, CONST_FLOAT_BITS_CONV, CONST_FLOAT_CLASSIFY, REPEAT_N }
```
[the remark this answers is not in the document]
Comment (rspencer01) [reply clippy_config/src/msrvs.rs]: Done, and changed test to reflect.

File: clippy_lints/src/matches/single_match.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (y21) [reply clippy_lints/src/matches/single_match.rs:254]: Ah, good point, I missed that it's not such a trivial change. I don't think it can implement Clone because it stores a mutable reference. Nevermind then.

File: clippy_lints/src/methods/mod.rs [hunk 13]
```diff
@@ -4166,6 +4167,41 @@ declare_clippy_lint! {
...
+    ///
+    /// ```no_run
+    /// let random_numbers : Vec<_> = (0..10).map(|_| { 3 + 1 }).collect();
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let f : Vec<_> = std::iter::repeat( 3 + 1 ).take(10).collect();
+    /// ```
+    ///
+    /// ### Known Issues
```
[the remark this answers is not in the document]
Comment (rspencer01) [reply clippy_lints/src/methods/mod.rs:4193]: Ah, I'd missed that in the notes. Does `Take<RepeatWith>` implement `DoubleEndedIterator` though? The playground (and release notes) suggest [not](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=96b643506efb391e70391de8a62e3d5f).

File: clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs (hunk 9 above)
Comment (rspencer01) [reply clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs:127]: I think it would be easier to just filter the vec. Let me know if you disagree stylistically.

File: clippy_lints/src/methods/mod.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (y21) [reply clippy_lints/src/methods/mod.rs:4193]: Right, it doesn't. I only actually looked at ExactSizeIterator and made wrong assumptions about DoubleEndedIterator

File: clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs (hunk 10 above)
Comment (rspencer01) [reply clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs:55]: Yes, this is probably an issue in that case. The applicability starts out as `MaybeIncorrect` I believe so doesn't need to be adjusted here though.

File: clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs (hunk 9 above)
Comment (y21) [reply clippy_lints/src/methods/map_with_unused_argument_over_ranges.rs]: I don't think the filtering is easier than something like
```rs
let mut parts = vec![
    (receiver.span.to(method_call_span), format!("std::iter::{method_to_use_name}")),
    new_span,
];
if use_take {
    parts.push((ex.span.shrink_to_hi(), format!(".take({count})")));
}
```

I also find it a lot clearer because the current filtering makes me think that there could be other replacements that are empty when it's (as far as I can tell?) really only the last one, and the last one can only be empty because of an explicit `String::new()` so it would be IMO clearer to just not create an empty replacement like that in the first place and only push a non-empty replacement if necessary.

(The debug assertion also only exists to find bugs in suggestions, so it feels more like a workaround)

**Comment by y21** [comment]
The lintcheck failure is weird, no idea what's up with that. Maybe try another rebase? Haven't seen this in other PRs

**Comment by rspencer01** [comment]
Tests seem to be passing. Waiting for a final OK then I'll squash and rebase.

**Comment by y21** [comment]
Looks all good to me now. Can you rebase one last time? Then I can r+

**Comment by y21** [comment]
Thanks!

@bors r+

**Comment by rspencer01** [comment]
Again, thank you @y21 for your guidance!
